cloudbridge/cloud/providers/aws/resources.py

```diff
@@ -788,91 +688,83 @@ def to_json(self):
+            'ToPort': self.to_port,
+            'IpRanges': [{'CidrIp': self.cidr_ip}] if self.cidr_ip else None,
+            'UserIdGroupPairs': [{
+                'GroupId': self.group_id}
+            ] if self.group_id else None
+        }
+
+        # Filter out empty values to please Boto
+        ip_perms = [trim_empty_params(ip_perm_entry)]
+
+        self.parent._security_group.revoke_ingress(IpPermissions=ip_perms)
+        self.parent._security_group.reload()
 
 
 class AWSBucketObject(BaseBucketObject):
 
-    def __init__(self, provider, key):
+    class BucketObjIterator():
+        CHUNK_SIZE = 4096
+
+        def __init__(self, body):
+            self.body = body
+
+        def __iter__(self):
+            while True:
+                data = self.read(self.CHUNK_SIZE)
+                if data:
+                    yield data
+                else:
+                    break
+
+        def read(self, length):
+            return self.body.read(amt=length)
+
+        def close(self):
+            return self.body.close()
+
+    def __init__(self, provider, obj):
         super(AWSBucketObject, self).__init__(provider)
-        self._key = key
+        self._obj = obj
 
     @property
     def id(self):
-        return self._key.name
+        return self._obj.key
 
     @property
     def name(self):
-        """
-        Get this object's name.
-        """
-        return self._key.name
+        return self.id
 
     @property
     def size(self):
-        """
-        Get this object's size.
-        """
-        return self._key.size
+        return self._obj.size
 
     @property
     def last_modified(self):
-        """
-        Get the date and time this object was last modified.
-        """
-        if self._key.last_modified:
-            lm = datetime.strptime(self._key.last_modified,
-                                   "%Y-%m-%dT%H:%M:%S.%fZ")
-            return lm.strftime("%Y-%m-%dT%H:%M:%S.%f")
-        else:
-            return None
+        return self._obj.last_modified.strftime("%Y-%m-%dT%H:%M:%S.%f")
 
     def iter_content(self):
-        """
-        Returns this object's content as an
-        iterable.
-        """
-        return self._key
+        return self.BucketObjIterator(self._obj.get().get('Body'))
 
     def upload(self, data):
-        """
-        Set the contents of this object to the data read from the source
-        string.
-        """
-        self._key.set_contents_from_string(data)
+        self._obj.put(Body=data)
 
     def upload_from_file(self, path):
-        """
-        Store the contents of the file pointed by the "path" variable.
-        """
-        self._key.set_contents_from_filename(path)
+        self._obj.upload_file(path)
 
     def delete(self):
-        """
-        Delete this object.
-
-        :rtype: bool
-        :return: True if successful
-        """
-        self._key.delete()
+        self._obj.delete()
 
     def generate_url(self, expires_in=0):
```

Comment:
Unless I'm looking at the wrong docs, the default expiry value is not valid: http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html

Looks like 1 is min and 604800 (seven days) is max?

Reply:
Looks like an error at the interface definition level. We should probably make expires_in default to None, and if it's specified, we need to mandate some default values that apply across all cloud providers.

Reply:
It doesn't look like there's max length of time before a URL expires on OpenStack (https://docs.openstack.org/swift/latest/api/temporary_url_middleware.html) so perhaps we should stick to the AWS 7 day max and default to 24 hours? Ping @VJalili

Reply:
I usually set it to 7200 min (and it is mainly because of large genomic files); but I guess 24h as default is reasonable.